### PR TITLE
TY: fix deref adjustments

### PR DESCRIPTION
**Pull request by vlad20012**
This shouldn't affect our users anyhow. Will be used by next PRs

### CODE REVIEW

File: src/main/kotlin/org/rust/lang/core/resolve/NameResolution.kt [hunk 1]
```diff
@@ -109,9 +109,16 @@ fun processFieldExprResolveVariants(
...
 ): Boolean {
-    for ((i, ty) in lookup.coercionSequence(receiverType).withIndex()) {
-        if (ty !is TyAdt || ty.item !is RsStructItem) continue
-        if (processFieldDeclarations(ty.item) { processor(FieldResolveVariant(it.name, it.element!!, ty, i)) }) return true
+    val derefSequence = mutableListOf<Ty>()
+    lookup.coercionSequence(receiverType).forEachIndexed { i, ty ->
+        if (i != 0) {
+            derefSequence += ty
+        }
+        if (ty !is TyAdt || ty.item !is RsStructItem) return@forEachIndexed
```
Comment (ortem) [inline src/main/kotlin/org/rust/lang/core/resolve/NameResolution.kt]: There was `continue` before, but now it is replaced with `return@forEachIndexed`, so semantics has changed. Is it correct?

File: src/main/kotlin/org/rust/lang/core/types/infer/TypeInference.kt [hunk 2]
```diff
@@ -32,19 +33,25 @@ fun inferTypesIn(element: RsInferenceContextOwner): RsInferenceResult {
...
+        override fun superVisitWith(visitor: TypeVisitor): Boolean = target.visitWith(visitor)
+    }
     class BorrowReference(
         override val target: TyReference,
-        val region: Region? = (target as? TyReference)?.region,
-        val mutability: Mutability? = (target as? TyReference)?.mutability
-    ) : Adjustment(target)
+        val region: Region? = target.region,
+        val mutability: Mutability? = target.mutability
+    ) : Adjustment(target){
```
Comment (ortem) [inline src/main/kotlin/org/rust/lang/core/types/infer/TypeInference.kt]:
```suggestion
    ) : Adjustment(target) {
```

File: src/main/kotlin/org/rust/lang/core/types/infer/TypeInference.kt [hunk 3]
```diff
@@ -32,19 +33,25 @@ fun inferTypesIn(element: RsInferenceContextOwner): RsInferenceResult {
         ?: error("Can not run nested type inference")
 }
 
-sealed class Adjustment(open val target: Ty) {
-    class Deref(target: Ty) : Adjustment(target)
+sealed class Adjustment(open val target: Ty): TypeFoldable<Adjustment> {
+    class Deref(target: Ty) : Adjustment(target) {
+        override fun superFoldWith(folder: TypeFolder): Adjustment = Deref(target.foldWith(folder))
+        override fun superVisitWith(visitor: TypeVisitor): Boolean = target.visitWith(visitor)
+    }
```
Comment (ortem) [inline src/main/kotlin/org/rust/lang/core/types/infer/TypeInference.kt]:
```suggestion
    }

```

File: src/main/kotlin/org/rust/lang/core/types/Extensions.kt [hunk 4]
```diff
@@ -85,6 +85,12 @@ val RsPatField.type: Ty
 val RsExpr.type: Ty
     get() = inference?.getExprType(this) ?: TyUnknown
 
+val RsExpr.adjustments: List<Adjustment>
```
[the remark this answers is not in the document]
Comment (vlad20012) [reply src/main/kotlin/org/rust/lang/core/types/Extensions.kt]: Used in #4768

File: src/main/kotlin/org/rust/lang/core/resolve/NameResolution.kt (hunk 1 above)
Comment (vlad20012) [reply src/main/kotlin/org/rust/lang/core/resolve/NameResolution.kt]: return in forEach is equivalent to continue in a loop

**Comment by vlad20012** [comment]
bors r=ortem

**Pull request merged**